### ISSUE TITLE
MlFlow Pipelines Transformer output 

### ISSUE THREAD

**Issue by luistelmocosta** [mlflow/mlflow#7146]
Hello! I have been trying the MlFlow Pipelines lately and am noticing that I don't know if it is working as intended. 
The transform step should use user-defined transformation functions and then apply them to the train and validation dataset from the split step.
However, when I use the transformation step to encode some of my features it only returns the features that are _transformed_.
I followed [this example](https://github.com/mlflow/mlp-regression-example/blob/master/steps/transform.py) and I understand why the author is only getting the feature he needs for training, but what happens when you want both, the non-transformed and transformed features?

Thank you!

**Comment by JHibbard** [comment]
Hi luistelmocosta,

A partial solution is to use sklearn's `sklearn.preprocessing.FunctionTransformer` with the `func` argument as None, this makes `func` the identity function. [Documentation here](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.FunctionTransformer.html#sklearn-preprocessing-functiontransformer).

```python
from sklearn.preprocessing import FunctionTransformer

# will return a column unchanged
FunctionTransformer()
```

I'll take a look if there's a better alternative to manually using identity functions, otherwise you may have found a good feature to add :)

Thank you,
Jim

**Comment by harupy** [comment]
@luistelmocosta Here's an example to keep non-trasformed features:

```python
import sklearn
from sklearn.feature_extraction import FeatureHasher
from sklearn.preprocessing import MinMaxScaler
from sklearn.compose import ColumnTransformer
import pandas as pd

print(sklearn.__version__)

X = pd.DataFrame(
    {
        "documents": ["First item", "second one here", "Is this the last?"],
        "width": [3, 4, 5],
    }
)
ct = ColumnTransformer([("num_preprocess", MinMaxScaler(), ["width"])], remainder="passthrough")

ct.fit(X)
print(ct.transform(X))
print(ct.get_feature_names_out())
```

Output:

```
1.1.2
[[0.0 'First item']
 [0.5 'second one here']
 [1.0 'Is this the last?']]
['num_preprocess__width' 'remainder__documents']
```

**Comment by luistelmocosta** [comment]
Thank you for your replies. The documentation says the transform step should define a `fit` and `transform` method. 
I included some feature engineering before the scaling transformation using a  `FunctionTransformer`. However, running the scale after that makes the pipeline step return only the scaled features. 

`"""
This module defines the following routines used by the 'transform' step of the regression pipeline:
- ``transformer_fn``: Defines customizable logic for transforming input data before it is passed
  to the estimator during model inference.
"""
import sklearn
import datetime
from pandas import DataFrame
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler, RobustScaler, MinMaxScaler, LabelEncoder, FunctionTransformer
import numpy as np

#Features selected based on Boruta (from another notebook)

COLS_SELECTED = [
    'store',
    'promo',
    'store_type',
    'assortment',
    'competition_distance',
    'competition_open_since_month',
    'competition_open_since_year',
    'promo2',
    'promo2_since_week',
    'promo2_since_year',
    'competition_time_month',
    'promo_time_week',
    'day_of_week_sin',
    'day_of_week_cos',
    'month_sin',
    'month_cos',
    'day_sin',
    'day_cos',
    'week_of_year_sin',
    'week_of_year_cos']


def calculate_features(df: DataFrame):
  # year week
  df['year_week'] = df['date'].dt.strftime( '%Y-%W' )

  # competition since
  df['competition_since'] = df.apply( lambda x: datetime.datetime( year=x['competition_open_since_year'], month=x['competition_open_since_month'],day=1 ), axis=1 )
  df['competition_time_month'] = ( ( df['date'] - df['competition_since'] )/30 ).apply( lambda x: x.days ).astype( int )

  # promo since
  df['promo_since'] = df['promo2_since_year'].astype( str ) + '-' + df['promo2_since_week'].astype( str )
  df['promo_since'] = df['promo_since'].apply( lambda x: datetime.datetime.strptime( x + '-1', '%Y-%W-%w' ) - datetime.timedelta( days=7 ) )
  df['promo_time_week'] = ( ( df['date'] - df['promo_since'] )/7 ).apply( lambda x: x.days ).astype( int )

  # assortment
  df['assortment'] = df['assortment'].apply( lambda x: 'basic' if x == 'a' else 'extra' if x == 'b' else 'extended' )

  # state holiday
  df['state_holiday'] = df['state_holiday'].apply( lambda x: 'public_holiday' if x == 'a' else 'easter_holiday' if x == 'b' else 'christmas' if x == 'c' else 'regular_day' )

  ## Column Selection
  #cols_drop = ['open', 'promo_interval', 'month_map']
  #df = df.drop( cols_drop, axis=1 )

  ### Encoding
  le = LabelEncoder()
  df['store_type'] = le.fit_transform( df['store_type'] )

  # assortment - Ordinal Encoding
  assortment_dict = {'basic': 1,  'extra': 2, 'extended': 3}
  df['assortment'] = df['assortment'].map( assortment_dict )

  ### Nature Transformation

  # day of week
  df['day_of_week_sin'] = df['day_of_week'].apply( lambda x: np.sin( x * ( 2. * np.pi/7 ) ) )
  df['day_of_week_cos'] = df['day_of_week'].apply( lambda x: np.cos( x * ( 2. * np.pi/7 ) ) )

  # month
  df['month_sin'] = df['month_number'].apply( lambda x: np.sin( x * ( 2. * np.pi/12 ) ) )
  df['month_cos'] = df['month_number'].apply( lambda x: np.cos( x * ( 2. * np.pi/12 ) ) )

  # day 
  df['day_sin'] = df['day_number'].apply( lambda x: np.sin( x * ( 2. * np.pi/30 ) ) )
  df['day_cos'] = df['day_number'].apply( lambda x: np.cos( x * ( 2. * np.pi/30 ) ) )

  # week of year
  df['week_of_year_sin'] = df['week_of_year_number'].apply( lambda x: np.sin( x * ( 2. * np.pi/52 ) ) )
  df['week_of_year_cos'] = df['week_of_year_number'].apply( lambda x: np.cos( x * ( 2. * np.pi/52 ) ) )
    
  return df[COLS_SELECTED]


def transformer_fn():
    """
    Returns an *unfitted* transformer that defines ``fit()`` and ``transform()`` methods.
    The transformer's input and output signatures should be compatible with scikit-learn
    transformers.
    """
    
    
    function_transformer_params = (
        {}
        if sklearn.__version__.startswith("1.0")
        else {"feature_names_out": "one-to-one"}
    )
    
    return Pipeline(
        steps=[
           (
                "calculate_features",
                FunctionTransformer(calculate_features, **function_transformer_params),
            ),
           (
                "encoder",
                ColumnTransformer(
                    transformers=[
                        (
                            "robust_scaler",
                            RobustScaler(),
                            ["competition_distance", "competition_time_month"],
                        ),
                        (
                            "min_max_scaler",
                            MinMaxScaler(),
                            ["promo_time_week"],
                        )
                    ]
                ),
            ),
          
        ]
    )`

**Comment by harupy** [comment]
@luistelmocosta I think it's because you don't set `remainder="passthrough"` in `ColumnTransformer`.

**Comment by harupy** [comment]
Example:

```python
import sklearn
from sklearn.feature_extraction import FeatureHasher
from sklearn.preprocessing import MinMaxScaler
from sklearn.compose import ColumnTransformer
import pandas as pd

print(sklearn.__version__)

X = pd.DataFrame(
    {
        "documents": ["First item", "second one here", "Is this the last?"],
        "width": [3, 4, 5],
    }
)
ct1 = ColumnTransformer([("num_preprocess", MinMaxScaler(), ["width"])], remainder="passthrough")
ct2 = ColumnTransformer([("num_preprocess", MinMaxScaler(), ["width"])])

print(ct1.fit_transform(X))
print(ct2.fit_transform(X))
```

```
[[0.0 'First item']
 [0.5 'second one here']
 [1.0 'Is this the last?']]
[[0. ]
 [0.5]
 [1. ]]
```

**Comment by MariaSdSantos** [comment]
Hello! I'm also working on this project and our problem is that we can get all the features with the remainder="passthrough", however, the names of the columns in the table get changed to this:
<img width="61" alt="image" src="https://user-images.githubusercontent.com/101735605/197823298-aa4edd2d-bc21-4d5f-b111-68c777e31a17.png">
How can we keep the original column names?

**Comment by harupy** [comment]
@MariaSdSantos MLflow Pipeline attempts to get the output feature names using `get_feature_names_out`. If this fails, mlflow falls back to auto-generated featue names.

https://github.com/mlflow/mlflow/blob/7802f5a4b5d114dbf6a9816515a856d0d8f489a9/mlflow/pipelines/steps/transform.py#L40

**Comment by harupy** [comment]
@luistelmocosta Please us know if `remainder="passthrough"` doesn't help.

**Comment by MariaSdSantos** [comment]
Hello. Thank you for your reply. `remainder="passthrough"` did help but the output column names do not match the original ones. Instead, it is adding the step name as a prefix and that's why we cannot get the correct names.
```
pipeline = Pipeline(
        steps=[
           (
                "calculate_features",
                FunctionTransformer(calculate_features, **function_transformer_params),
            ),
          (
                "encoder",
                ColumnTransformer(
                    transformers=[
                        (
                            "robust_scaler",
                            RobustScaler(),
                            ["competition_distance", "competition_time_month"],
                        ),
                        (
                            "min_max_scaler",
                            MinMaxScaler(),
                            ["promo_time_week"],
                        )
                    ], remainder="passthrough"
                ),
            ),
          
        ]
    ) 
```
```
function_transformer_params = (
        {}
        if sklearn.__version__.startswith("1.0")
        else {"feature_names_out": ""}
    )
```

```
pipeline.fit_transform(training_data)
pipeline.get_feature_names_out(COLS_SELECTED)
```
<img width="291" alt="image" src="https://user-images.githubusercontent.com/101735605/198059629-b2dfe529-2fcd-4513-86e3-894a375b0f16.png">


The `calculate_features ` function is the same as the above. Since there is no match, it is returning those f_* column names. Thank you

**Comment by harupy** [comment]
Can you add `FunctionTransformer` that strips the `remainder__` prefix?

**Comment by MariaSdSantos** [comment]
Hello. We tried to add strip to the pipeline. However, we are not sure where to apply it in the pipeline, because the output of the transformers are an array and not a dataframe.
```
def strip_columns(df: DataFrame):
    df.columns = [x.split('__')[-1] for x in df.columns]
    return df

pipeline = Pipeline(
        steps=[
           (
                "calculate_features",
                FunctionTransformer(calculate_features, **function_transformer_params),
            ),
          (
                "encoder",
                ColumnTransformer(
                    transformers=[
                        (
                            "robust_scaler",
                            RobustScaler(),
                            ["competition_distance", "competition_time_month"],
                        ),
                        (
                            "min_max_scaler",
                            MinMaxScaler(),
                            ["promo_time_week"],
                        )
                    ], remainder="passthrough"
                ),
            ),
            (
                "rename_features",
                FunctionTransformer(strip_columns, **function_transformer_params),
            ),
          
        ]
    )
```
Thank you.

**Comment by harupy** [comment]
https://github.com/scikit-learn/scikit-learn/issues might be a better place to ask this.
